### ISSUE TITLE
Method mapping is failing for a simple single field scenario

### ISSUE THREAD

**Issue by jenseralmeida** [JonPSmith/EfCore.GenericServices#31]
When calling the code below the following exception is triggered

``` C#
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using GenericServices;
using GenericServices.PublicButHidden;
using GenericServices.Setup;
using Microsoft.EntityFrameworkCore;
using Xunit;

namespace EfCore.GenericServices.Testing
{
    public class SingleFieldMethodMappingTests: IDisposable
    {
        private readonly SQLiteInMemoryOptions<RootContext> _options = new SQLiteInMemoryOptions<RootContext>();

        [Fact]
        public async Task Root_method_call_fails()
        {
            // arrange
            const string item = @"A item!";

            using (var context = new RootContext(_options))
            {
                var utData = context.SetupSingleDtoAndEntities<SetItemDto>();
                var service = new CrudServicesAsync(context, utData.ConfigAndMapper);

                // act
                var dto = new SetItemDto
                {
                    Id = 1,
                    Item = item
                };
                await service.UpdateAndSaveAsync(dto);//, nameof(Root.SetItem)); // DEBUG: Exceptions is thrown. It has the same result when excplicitly setting the method name or not
            }
        }

        public void Dispose() => _options?.Dispose();

        public class SetItemDto : ILinkToEntity<Root>
        {
            [ReadOnly(true)]
            public int Id { get; set; }
            public string Item { get; set; }
        }

        public class Root
        {
            public int Id { get; set; }
            public void SetItem(string item, RootContext context = null)
            {
                if (item is null)
                    throw new ArgumentNullException(nameof(item));
                // This method is never called
            }
        }

        public class RootContext : DbContext
        {
            public RootContext(DbContextOptions<RootContext> options) : base(options) { }
            protected override void OnModelCreating(ModelBuilder builder)
            {
                builder.Entity<Root>().Property(e => e.Id).ValueGeneratedNever();
                builder.Entity<Root>().HasData(new Root { Id = 1 });
            }
        }
    }
}

``` 

```
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.0 (64-bit .NET Core 4.6.27617.05)
[xUnit.net 00:00:00.38]   Discovering: readonly-property
[xUnit.net 00:00:00.43]   Discovered:  readonly-property
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.0 (64-bit .NET Core 4.6.27617.05)
[xUnit.net 00:00:00.43]   Starting:    readonly-property
[xUnit.net 00:00:01.82]     readonly_property.RootMethodTest.Root_method_call_fails [FAIL]
[xUnit.net 00:00:01.82]       Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : Delegate 'System.Action<readonly_property.RootMethodTest.SetItemDto,readonly_property.RootMethodTest.Root,readonly_property.RootMethodTest.RootContext>' has some invalid arguments
[xUnit.net 00:00:01.82]       Stack Trace:
[xUnit.net 00:00:01.82]            at CallSite.Target(Closure , CallSite , Object , Object , Object , DbContext )
[xUnit.net 00:00:01.82]            at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid4[T0,T1,T2,T3](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
[xUnit.net 00:00:01.82]            at GenericServices.Internal.LinqBuilders.BuildCall.RunMethodViaLinq(MethodInfo methodInfo, Object dto, Object entity, List`1 propertyMatches, DbContext context)
[xUnit.net 00:00:01.82]            at System.Dynamic.UpdateDelegates.UpdateAndExecute6[T0,T1,T2,T3,T4,T5,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5)
[xUnit.net 00:00:01.82]            at GenericServices.Internal.MappingCode.EntityUpdateHandler`1.ReadEntityAndUpdateViaDto(TDto dto, String methodName)
[xUnit.net 00:00:01.82]            at GenericServices.PublicButHidden.CrudServicesAsync`1.UpdateAndSaveAsync[T](T entityOrDto, String methodName)
[xUnit.net 00:00:01.82]         c:\src\local\readonly-property\DddTest copy.cs(33,0): at readonly_property.RootMethodTest.Root_method_call_fails()
[xUnit.net 00:00:01.82]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:00:01.83]   Finished:    readonly-property
----- Test Execution Summary -----

readonly_property.RootMethodTest.Root_method_call_fails:
    Outcome: Failed
    Error Message:
    Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : Delegate 'System.Action<readonly_property.RootMethodTest.SetItemDto,readonly_property.RootMethodTest.Root,readonly_property.RootMethodTest.RootContext>' has some invalid arguments
    Stack Trace:
       at CallSite.Target(Closure , CallSite , Object , Object , Object , DbContext )
   at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid4[T0,T1,T2,T3](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at GenericServices.Internal.LinqBuilders.BuildCall.RunMethodViaLinq(MethodInfo methodInfo, Object dto, Object entity, List`1 propertyMatches, DbContext context)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute6[T0,T1,T2,T3,T4,T5,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5)
   at GenericServices.Internal.MappingCode.EntityUpdateHandler`1.ReadEntityAndUpdateViaDto(TDto dto, String methodName)
   at GenericServices.PublicButHidden.CrudServicesAsync`1.UpdateAndSaveAsync[T](T entityOrDto, String methodName)
   at readonly_property.RootMethodTest.Root_method_call_fails() in c:\src\local\readonly-property\DddTest copy.cs:line 33
--- End of stack trace from previous location where exception was thrown ---
    
Total tests: 1. Passed: 0. Failed: 1. Skipped: 0
```

The complete solution code can be seen at https://gist.github.com/jenseralmeida/b299a11de8e5b281e7af39e83838ae4c

**Comment by JonPSmith** [comment]
Hi @jenseralmeida 

I can see the error but I don't think your PR wouldn't cover all the usages of the `ParametersMatch` class. Have you run all the unit tests on you version? Do they all pass with your changes?

Can I understand what you are wanting to do. From your original issue #30 I assume you want to have DTOs with non-public, or missing, setters. I can see that would make sense in some cases, and would be the equivalent of adding `[ReadOnly(true)]`. Is that what you want? If I have time I could look at that.

Let me know.

**Comment by jenseralmeida** [comment]
Hi @JonPSmith ,

Thanks for taking the time to evaluate this. So, this issue is unrelated to issue #30, and I will be answering the points related to the PR and issue #30 on the corresponding issue.

Now, getting back to this issue, the code presented on it is a simplification of one code I have tried and the described exception happened. For this issue, I could not understand how to work around the error and neither I could not find anything wrong with my DTO. 

(again, this is different from issue #30 where I did try to use an invalid DTO, and I will be discussing the PR in there)

**Comment by JonPSmith** [comment]
OK, Now I understand a bit more. 

Issue #31 you have hit a bug that I know about but haven't got round to fixing it because a) its complicated and b) there is a workaround, c) I was busy!

I have a fix now and working on #30 before I issue a new version. My be a delay as I have client work coming up.

**Comment by JonPSmith** [comment]
This is fixed in version 2.0.3.
